### PR TITLE
SALTO-1085: omit hidden annotation values in deleted fields

### PR DESCRIPTION
**Pull request by ori-moisis**
Generally mergeWithHidden is supposed to add hidden values to workspace elements but it should not do so if the parent of the hidden value was deleted in the workspace.
However this did not work with hidden annotation values in fields because we only filtered out top level elements that were deleted

---
_Release Notes_
- Fixed issue that caused restoring deleted fields from state to fail

### CODE REVIEW

File: packages/workspace/src/workspace/hidden_values.ts [hunk 1]
```diff
@@ -89,23 +90,30 @@ const splitElementHiddenParts = <T extends Element>(
...
     transformFunc: ({ value, path }) => (isPartOfHiddenPath(path) ? value : undefined),
     strict: true,
   })
   // remove all annotation types from the hidden element so they don't cause merge conflicts
   hidden.annotationTypes = {}
-  return { hidden, visible }
+  // Workaround for transformElement not being able to omit fields from an ObjectType
+  if (isObjectType(hidden) && isObjectType(workspaceElement)) {
+    // filter out fields that were deleted in the workspace
+    hidden.fields = _.pick(hidden.fields, Object.keys(workspaceElement.fields))
```
Comment (netama) [inline packages/workspace/src/workspace/hidden_values.ts:102]: I'm not sure I understand this one - won't it exclude hidden fields from `hidden`?

Comment (netama) [reply packages/workspace/src/workspace/hidden_values.ts:102]: I think this should also include fields with `_hidden = true` and then it will be ok - maybe something like
```suggestion
    hidden.fields = _.pickBy(hidden.fields, (field, name) => workspaceElement.fields[name] !== undefined && !isHidden(field))
```

Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:102]: as you said, it is not a real scenario because we don't support hiding field definitions.
you are correct that if we implement that feature without changing this it will break, not sure that means we need to change it now.

File: packages/workspace/src/workspace/hidden_values.ts [hunk 2]
```diff
@@ -37,12 +37,13 @@ const isHidden = (element?: Element): boolean => (
   || ((isInstanceElement(element) || isField(element)) && isHiddenValue(element.type))
 )
 
-const splitElementHiddenParts = <T extends Element>(
-  element: T,
-): { hidden?: T; visible?: T } => {
-  if (isHidden(element)) {
+const getElementHiddenParts = <T extends Element>(
```
Comment (netama) [inline packages/workspace/src/workspace/hidden_values.ts:40]: just to make sure I understand - the switch to just returning the hidden part is an unrelated optimization because `.visible` was never used, right?

Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:40]: yes
not really a performance optimization, more of a readability optimization

**Review by netama — APPROVED** [review]
one comment

File: packages/workspace/src/workspace/hidden_values.ts (hunk 1 above)
Comment (netama) [reply packages/workspace/src/workspace/hidden_values.ts:102]: :+1: added the details on https://salto-io.atlassian.net/browse/SALTO-1086 so if you prefer not to support that now hopefully we'll find it then

Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts]: changed
